### PR TITLE
Draft: Take screenshot when ammo counter goes down

### PR DESCRIPTION
**Pull request by MadLittleMods**
Follow-up from https://github.com/MadLittleMods/fps-aim-analyzer/pull/7

---

Because I tend to click a lot more than the gun can fire, taking a screenshot on mouse-click adds a lot of extra screenshots that don't matter. Additionally, there is a lot of input-lag to the game so even the screenshots that matter are pretty far off from where they end up hitting.

Instead with this PR, we take a screenshot whenever we see the ammo counter go down which should be an accurate time to when the bullet comes out of gun and take a screenshot of what we're aiming at.

For text binarization, TODO

For OCR, we're using my [neural network library](https://github.com/MadLittleMods/zig-neural-networks) to recognize the digits.

HUD opacity | On gameplay | On a white background (still gameplay)
--- | --- | ---
50% (default) |  | 
100% |  | 

32/72 ammo in BR75 (Halo Infinite gameplay)

Example tougher cases:

01 - cliffhanger red hex

13 - cliffhanger massive red blue hex

### Dev notes

 - Overlay text
 - UI text
 - Closed captions
 - Scene text recognition (OCR in natural scenes *which is not what we need*)
 - "graphic text" (digitally added as an overlay on videos or images) vs "scene text" (text on objects, captured in its native environment) ([mentioned in this *Text Recognition in the Wild: A Survey* paper](https://arxiv.org/pdf/2005.03492.pdf))
 - Text detection: Detect whether text is present
 - Text recognition: Get a bounding box of text
 - Text segmentation: Separate lines and characters of text

---

OCR preprocessing:

 - Image thresholding with OpenCV, https://docs.opencv.org/3.4/d7/d4d/tutorial_py_thresholding.html
 - [A Novel Approach for Binarization of Overlay Text](https://islab.ulsan.ac.kr/files/announcement/609/A%20novel%20approach%20for%20binarization%20of%20overlay%20text.pdf) (Zhike Zhang, Weiqiang Wang)
 - [Overlay Text Extraction From TV News Broadcast](https://arxiv.org/pdf/1604.00470.pdf) (Raghvendra Kannao, Prithwijit Guha)
 - [An Extraction of Overlay Text from Digital Videos](https://researchinventy.com/papers/v5i4/O054088094.pdf) (Dr. Saravanan, N Lavanya, K Mahesh)
 - [A New Approach for Overlay Text Detection and Extraction From Complex Video Scene](https://web.archive.org/web/20111206122119id_/http://cilabs.kaist.ac.kr/People/paper/TIP2009_WJ.pdf) (Wonjun Kim, Changick Kim)
 - [Overlay Text Graphic Region Extraction for Video Quality Enhancement Application](https://koreascience.kr/article/JAKO201322658550170.pdf) (Sanghee Lee, Hansung Park, Jungil Ahn, Youngsang On, and Kanghyun Jo)
 - [Accurate Overlay Text Extraction for Digital Video Analysis](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=abae08beec0e72fc70bce2425911f150d66d4a61) (Dongqing Zhang, Shih-Fu Chang)
 - [Color, Shape and Texture: Feature Extraction using OpenCV](https://web.archive.org/web/20220219085749/https://medium.com/mlearning-ai/color-shape-and-texture-feature-extraction-using-opencv-cb1feb2dbd73)

Fixing chromatic aberration:

 - Chromatic aberration correction algorithm, https://github.com/RayXie29/Chromatic_aberration_correction/tree/master
 - [Fast Chromatic Aberration Correction with 1D Filters](https://www.ipol.im/pub/art/2023/443/article.pdf) (Thomas Eboli)
 - Blind Chromatic Aberration Correction with False Color Filtering, https://github.com/teboli/chromatic_aberration_filtering

Text region detection:

 - https://stackoverflow.com/questions/24385714/detect-text-region-in-image-using-opencv/59977588#59977588

OpenCV:

 - https://pyimagesearch.com/2017/08/28/fast-optimized-for-pixel-loops-with-opencv-and-python/
 - https://stackoverflow.com/questions/10948589/choosing-the-correct-upper-and-lower-hsv-boundaries-for-color-detection-withcv?rq=3

Morphological image processing:

 - Erosion/Dilation and opening/closing: https://towardsdatascience.com/understanding-morphological-image-processing-and-its-operations-7bcf1ed11756
 - https://docs.opencv.org/4.8.0/d9/d61/tutorial_py_morphological_ops.html
 - https://blog.ostermiller.org/efficiently-implementing-dilate-and-erode-image-functions/

Contour detection:

 - Great guide for square, moore neighbor, radial sweep, Theo Pavlidis countour boundary tracing algorithms, https://www.imageprocessingplace.com/downloads_V3/root_downloads/tutorials/contour_tracing_Abeer_George_Ghuneim/square.html
 - https://theailearner.com/tag/contour-tracing-algorithms/
 - Suzuki algorithm implementation (what OpenCV uses) https://github.com/RajdeepMondal/Contour-Detection
 - https://www.eriksmistad.no/moore-neighbor-contour-tracing-algorithm-in-c/
 - https://gist.github.com/cwellsx/e9a1d7092203073c40565455c9b5c79d
 - https://github.com/helviett/Diploma/blob/6e73fdbba49eb03d172c5382c7e1fdc53c55d6b8/Bergamot/NonConvexHullGeneration/NonConvexHull.cs

Dealing with PNG's in Zig:

 - https://github.com/zigimg/zigimg
 - https://igor84.github.io/blog/png-reader-in-zig/

Resizing images (interpolation, sampling):

 - Nearest neighbor, bilinear, bicubic, https://blog.demofox.org/2015/08/15/resizing-images-with-bicubic-interpolation/
 - https://blog.codinghorror.com/better-image-resizing/
 - https://matplotlib.org/stable/gallery/images_contours_and_fields/interpolation_methods.html
 - https://theailearner.com/2018/12/29/image-processing-bilinear-interpolation/
 - https://theailearner.com/2018/12/29/image-processing-bicubic-interpolation/
 - https://bartwronski.com/2021/02/15/bilinear-down-upsampling-pixel-grids-and-that-half-pixel-offset/
 - http://alvyray.com/Memos/CG/Microsoft/6_pixel.pdf
 - https://iquilezles.org/articles/hwinterpolation/
 - https://magcius.github.io/xplain/article/rast1.html#sampling
 - [Bilinear implementation in SerenityOS](https://github.com/SerenityOS/serenity/blob/09124fc3a52262eb45f556c69786132b685cb7fe/Userland/Libraries/LibGfx/Painter.cpp#L1228-L1248)

### CODE REVIEW

File: src/vision/image_conversion.zig [hunk 1]
```diff
@@ -0,0 +1,1037 @@
...
+    // Interpolate bi-linearly!
+    const PixelType = std.meta.Child(@TypeOf(source_image.pixels));
+    switch (PixelType) {
+        RGBPixel => {
+            return .{
+                .r = bilerp(p00.r, p10.r, p01.r, p11.r, x_fractional, y_fractional),
+                .g = bilerp(p00.g, p10.g, p01.g, p11.g, x_fractional, y_fractional),
+                .b = bilerp(p00.b, p10.b, p01.b, p11.b, x_fractional, y_fractional),
+            };
+        },
```
Comment (MadLittleMods) [inline src/vision/image_conversion.zig]: For some reason, this gives different results than what I see in other applications. Haven't nailed this down yet :thinking: 

### Our implementation

```
zig build test -freference-trace --verbose --summary all -Dtest-filter="resizeImage"
```

16x16 test square bilinear downscale to 4x4:

test-square






8x8 mooshroom cow bilinear upscale 32x32:

(it is a bit different but we don't have any shifting or edge artifact problems)

cow-mooshroom

2024-01-25_07-43 - ours vs chrome mooshroom cow



### Benchmark

Comparing this to the `PIL.image.resize` result from https://medium.com/hackernoon/how-tensorflows-tf-image-resize-stole-60-days-of-my-life-aba5eb093f35, ours is different.


Also tested this result myself:
```py
from PIL import Image

with Image.open("images/test-square.png") as im:
    new_width = 4
    new_height = 4
    im_resized = im.resize((new_width, new_height), Image.Resampling.BILINEAR)
    im.show()
    im_resized.show()
```

PIL resized image

Chrome (25% zoom):

Chrome at 25% zoom

Chrome (resized ``):

Chrome img tag



Firefox (resized ``):

Firefox img tag


Ladybird (resized ``):

Ladybird img tag




Gimp (linear resize):

Gimp linear resize

Comment (MadLittleMods) [reply src/vision/image_conversion.zig]: I made things better by iterating the uv in the range `[0.5/width, 1 – 0.5/width]` (https://github.com/MadLittleMods/fps-aim-analyzer/pull/8/commits/1fa7179a48f2c536be15f82101a461db06c8bf37) thanks to this comment:

> If you have a 1D texture of size 4, what are your pixel/texel coordinates?
> 
> They can be [0, 1, 2, 3]. But GPUs use a convention of half pixel offsets, so they end up being [0.5, 1.5, 2.5, 3.5]. This translates to UVs, or “normalized” coordinates [0.5/4, 1.5/4, 2.5/4, 3.5/4], which spans a range of [0.5/width, 1 – 0.5/width].
> 
> This representation seems counterintuitive at first, but what it provides us is a guarantee and convention that the image corners are placed at [0 and 1] normalized, or [0, width] unnormalized.
>
> *-- https://bartwronski.com/2021/02/15/bilinear-down-upsampling-pixel-grids-and-that-half-pixel-offset/*



This made things "better" but still not the same as these other applications. It's obvious to see that it isn't taking the blue into account to make purple and even from the math it's obvious. I'm not sure how these other applications ending up with that result using "standard" bilinear.

Comment (MadLittleMods) [reply src/vision/image_conversion.zig]: I added a new interpolation resize method that just averages all pixels in the UV region for the new pixel `sampleAverage` (https://github.com/MadLittleMods/fps-aim-analyzer/pull/8/commits/001ca05feb3237b98111e9f1443fbf3d163031ca) which results the ~exact same image that Chrome gives. Would be good to find the code that is responsible for this behavior.

Comment (MadLittleMods) [reply src/vision/image_conversion.zig]: The logic for the Ladybird browser (Serenity) is in [`to_gfx_scaling_mode(...)` -> `Userland/Libraries/LibWeb/CSS/ComputedValues.h#L270-L285`](https://github.com/SerenityOS/serenity/blob/99fbd33d7d887723372e861be27f2743d335ecb5/Userland/Libraries/LibWeb/CSS/ComputedValues.h#L270-L285). If the source is bigger than what you're trying to display, it does box sampling (averaging everything), otherwise bilinear

File: src/vision/morphological_operations.zig [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+const BinaryImage = image_conversion.BinaryImage;
+const BinaryPixel = image_conversion.BinaryPixel;
+const binaryPixelsfromIntArray = image_conversion.binaryPixelsfromIntArray;
+const binaryToRgbImage = image_conversion.binaryToRgbImage;
+const expectImageEqual = image_conversion.expectImageEqual;
+const math_utils = @import("../utils/math_utils.zig");
+const absoluteDifference = math_utils.absoluteDifference;
+const print_utils = @import("../utils/print_utils.zig");
+const printLabeledImage = print_utils.printLabeledImage;
+
```
Comment (MadLittleMods) [inline src/vision/morphological_operations.zig:14]: Nowadays, if we wanted to use a library, [`zignal`](https://github.com/arrufat/zignal) might fit the bill
